change: Better configuration handling

Leo40Git opened the pull request:
- Corrupt config files will be replaced with a copy of the default values.
  - The corrupt file will be backed up for inspection.
  - The game will still crash, just to give people a heads-up that their config's hosed and has been replaced.
- Changed the way in-game options are saved to the file. This *should* fix the "corrupt config" bug.
  - More specifically, the new file is written to a temporary file, and then that temporary file is moved into the original file's place.
- The `core` mixin package can no longer be disabled.
- I/O now uses NIO `Path`s, because they're cool

Leo40Git committed:
change: Better configuration error handling

A corrupt config file will no longer bring the entire thing down!
Instead, the corrupt file will be backed up, and a copy of the default values will take its place.
(also I made everything use Path)

Leo40Git committed:
Use the normal logger

Leo40Git committed:
Restore "this."es

Leo40Git committed:
Correct some control flow hiccups

FlashyReese commented on src/main/java/me/jellysquid/mods/sodium/client/gui/SodiumGameOptions.java:
```diff
@@ -139,22 +144,38 @@ public String getLocalizedName() {
 
-        config.file = file;
+        config.path = path;
         config.writeChanges();
```
I have mentioned this in `#sodium` before. Perhaps we should only write changes once when creating the config file for this first time.
```suggestion
        if (!File.exists(path)) {
                config.writeChanges();
        }
```

Leo40Git committed:
SodiumGameOptions: only write changes if we've loaded default values

This prevents an unnecessary write if everything was loaded correctly.

Leo40Git committed:
SodiumGameOptions: Use atomic strategy to prevent config corruption

Write to temp file -> delete original file -> move temp file to original file's place using an atomic move operation

SodiumConfig: File is now backed up if it couldn't be read, *not* if it doesn't exist in the first place

Leo40Git committed:
oops, remove unused imports

Leo40Git committed:
SodiumGameOptions: crash on config read failure

Still gives people a heads-up that their config is hosed without breaking the game forever (until they manually replace the config)

Leo40Git committed:
SodiumConfig: write default config if we failed to read it

Leo40Git committed:
SodiumConfig: make load(Path) less horrible to look at

Leo40Git committed:
SodiumConfig: Add forced option support

No more disabling "core"!

Leo40Git committed:
SodiumGameOptions: Remove ATOMIC_MOVE, use REPLACE_EXISTING instead

PathUtil: Add some (probably unintelligible) javadoc to resolveTimestampedSibling

Leo40Git committed:
forgot a quote here

magneticflux- commented on src/main/java/me/jellysquid/mods/sodium/common/util/PathUtil.java:
```diff
@@ -0,0 +1,21 @@
+        LocalDateTime now = LocalDateTime.now();
+        String backupName = base.getFileName().toString();
+        backupName = backupName.substring(0, backupName.lastIndexOf('.') - 1);
+        backupName += "_" + tag + "_" + now.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME) + ".json";
```
This causes issues with path resolution because `:` is not a valid file name. Reported on the Discord here: https://discord.com/channels/602796788608401408/778685768041627719/819478369418215424

Leo40Git replied:
great job, past me. what do I even pay you for

Leo40Git replied:
Should be fixed in https://github.com/CaffeineMC/sodium-fabric/pull/571/commits/220c417fca9ae23a142661874879a31ee300cab5.

magneticflux- commented on src/main/java/me/jellysquid/mods/sodium/common/util/PathUtil.java:
```diff
@@ -0,0 +1,21 @@
+    public static Path resolveTimestampedSibling(Path base, String tag) {
+        LocalDateTime now = LocalDateTime.now();
+        String backupName = base.getFileName().toString();
+        backupName = backupName.substring(0, backupName.lastIndexOf('.') - 1);
```
~This is not a reliable way to determine the file name without extension. In fact, this website's first example of how not to ask a technical question uses file extension handling to convey the point: https://xyproblem.info/~

Nevermind, it appears the Microsoft .NET standard library uses the "last `.` truncation" technique: [https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getfilenamewithoutextension?redirectedfrom=MSDN&view=net-5.0#System_IO_Path_GetFileNameWithoutExtension_System_String_](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getfilenamewithoutextension?redirectedfrom=MSDN&view=net-5.0#System_IO_Path_GetFileNameWithoutExtension_System_String_)

magneticflux- replied:
@Leo40Git Hey, turns out a relevant PR in OpenJDK was opened recently! https://github.com/openjdk/jdk/pull/2319

Leo40Git committed:
oops: Remove colons from timestamped file name

Leo40Git committed:
change: Remove PathUtil.resolveTimestampedSibling's file extension handling

Change timestamp format to be more compact

Leo40Git committed:
change: Make mixin config also use "write to temp, move over original" technique
